File: backend/outbound.py
# =====================================================
# outbound.py - 불출 관련 API
# =====================================================
# 역할: 창고에서 물품을 꺼낼 때 처리하는 기능
#
# API 목록:
#   GET    /api/fifo?code=품번  -> 선입선출 순서 조회
#   POST   /api/outbound        -> 불출 처리 (재고 자동 차감)
#   DELETE /api/outbound/<id>   -> 불출 취소 (재고 자동 복원)
#   GET    /api/history         -> 불출 이력 조회
#
# 선입선출(FIFO) 규칙:
#   1순위: 환입 항목 (생산 라인에서 돌아온 것)
#   2순위: 입고 날짜 빠른 순
# =====================================================

from flask import Blueprint, request, jsonify
from database import load_data, save_data, add_record
from datetime import date
import time

outbound_bp = Blueprint('outbound', __name__)


@outbound_bp.route('/api/fifo', methods=['GET'])
def get_fifo():
    """
    선입선출(FIFO) 불출 순서 조회 API

    쿼리 파라미터:
        code : 품번 (끝 자리 검색 가능, 예: '0017')

    반환:
        해당 품번의 입고/환입 항목을 불출 우선순위 순으로 반환
        - 환입 항목이 무조건 1순위
        - 그 다음 입고 날짜 빠른 순
        - 소진된 항목(수량 0)은 제외
    """
    code = request.args.get('code', '').strip()
    if not code:
        return jsonify({'success': False, 'message': '품번을 입력하세요'}), 400

    today = date.today().isoformat()
    data = load_data()

    # 해당 품번 입고/환입 항목 필터링
    # - 소진되지 않은 것만
    # - 오늘 소진된 것은 포함 (당일은 소진 표시로 보임)
    matches = [
        d for d in data
        if d.get('kind') in ('in', 'hwanjip')
        and (d.get('code') or '').upper().endswith(code.upper())
        and not d.get('depleted')
        and ((d.get('qty') or 0) > 0 or (d.get('rolls') or 0) > 0)
    ]

    # 정렬: 환입 1순위, 그 다음 입고일 오름차순
    matches.sort(key=lambda x: (
        0 if x.get('kind') == 'hwanjip' else 1,  # 환입=0, 입고=1
        x.get('date', '')                          # 날짜 빠른 순
    ))

    return jsonify({'success': True, 'data': matches})


@outbound_bp.route('/api/outbound', methods=['POST'])
def add_outbound():
    """
    불출 처리 API

    받는 데이터 (JSON):
        code      : 품번
        name      : 품목명
        item_type : 품목 유형 ('normal' 또는 'fabric')
        person    : 담당자 이름 (필수!)
        from_id   : 차감할 입고 항목의 ID
        from_loc  : 불출 위치 (이력에 표시됨)
        qty       : 불출 수량 (일반)
        rolls     : 불출 롤 수 (원단)
        weight    : 불출 무게 (원단)
        date      : 불출 날짜
        note      : 비고

    처리 내용:
        1. from_id 항목에서 수량 차감
        2. 수량 0 되면 depleted=True (소진 처리)
        3. 불출 이력 저장
    """
    body = request.get_json()
    if not body:
        return jsonify({'success': False, 'message': '데이터가 없습니다'}), 400
    if not body.get('code'):
        return jsonify({'success': False, 'message': '품번을 입력하세요'}), 400
    if not body.get('person'):
        return jsonify({'success': False, 'message': '담당자 이름을 입력하세요'}), 400

    data = load_data()
    from_id = body.get('from_id')

    # 선택한 입고 항목에서 수량 차감
    if from_id:
        for item in data:
            if item.get('id') == from_id:
                if body.get('item_type') == 'fabric':
                    # 원단: 롤수와 무게 차감
                    item['rolls']  = max(0, (item.get('rolls')  or 0) - (body.get('rolls')  or 0))
                    item['weight'] = max(0, (item.get('weight') or 0) - (body.get('weight') or 0))
                    item['qty']    = item['rolls']
                else:
                    # 일반: 수량 차감
                    item['qty'] = max(0, (item.get('qty') or 0) - (body.get('qty') or 0))

                # 수량이 0이 되면 소진 처리
                if item.get('qty', 0) <= 0:
                    item['depleted'] = True
                break

    # 불출 이력 저장
    body['kind'] = 'out'
    body['id']   = int(time.time() * 1000)
    data.append(body)
    save_data(data)

    return jsonify({'success': True, 'data': body}), 201


@outbound_bp.route('/api/outbound/<int:record_id>', methods=['DELETE'])
def cancel_outbound(record_id):
    """
    불출 취소 API

    처리 내용:
        1. 불출 이력 삭제
        2. 원본 입고 항목 수량 복원
        3. 소진 해제 (depleted=False)
    
    사용 시점:
        생산라인에서 잘못 주문했을 때 취소
    """
    data = load_data()

    # 취소할 불출 이력 찾기
    out = next((d for d in data if d.get('id') == record_id and d.get('kind') == 'out'), None)
    if not out:
        return jsonify({'success': False, 'message': '불출 이력을 찾을 수 없습니다'}), 404

    # 원본 입고 항목 수량 복원
    from_id = out.get('from_id')
    if from_id:
        for item in data:
            if item.get('id') == from_id:
                if out.get('item_type') == 'fabric':
                    item['rolls']  = (item.get('rolls')  or 0) + (out.get('rolls')  or 0)
                    item['weight'] = (item.get('weight') or 0) + (out.get('weight') or 0)
                    item['qty']    = item['rolls']
                else:
                    item['qty'] = (item.get('qty') or 0) + (out.get('qty') or 0)
                item['depleted'] = False  # 소진 해제
                break

    # 불출 이력 삭제
    new_data = [d for d in data if d.get('id') != record_id]
    save_data(new_data)

    return jsonify({'success': True, 'message': '불출 취소 완료. 재고가 복원됐습니다'})


@outbound_bp.route('/api/history', methods=['GET'])
def get_history():
    """
    불출 이력 조회 API

    쿼리 파라미터:
        from : 시작 날짜 (YYYY-MM-DD)
        to   : 종료 날짜 (YYYY-MM-DD)

    반환:
        불출 기록 (최신순)
        - 날짜 필터 적용
        - 3개월(90일) 이전 이력 자동 삭제
        - 통계: 전체 건수, 필터 적용 건수
    """
    from datetime import timedelta

    today    = date.today()
    today_str = today.isoformat()

    # 3개월(90일) 이전 이력 자동 삭제
    cutoff = (today - timedelta(days=90)).isoformat()
    data   = load_data()
    before = len(data)
    data   = [d for d in data if not (d.get('kind') == 'out' and d.get('date', '') < cutoff)]
    if len(data) < before:
        save_data(data)

    # 불출 항목만 필터링
    outs = [d for d in data if d.get('kind') == 'out']
    total = len(outs)

    # 날짜 필터 적용
    date_filter = request.args.get('date', '')
    kind_filter = request.args.get('kind', 'out')  # 기본값은 'out'

    # 종류 필터 (입고/불출/환입)
    outs = [d for d in data if d.get('kind') == kind_filter]

    if date_filter:
        outs = [o for o in outs if o.get('date', '') == date_filter]

    # 최신순 정렬
    outs = sorted(outs, key=lambda x: x.get('date', ''), reverse=True)

    return jsonify({
        'success' : True,
        'data'    : outs,
        'total'   : total,           # 전체 불출 건수 (필터 없이)
        'filtered': len(outs),       # 필터 적용 후 건수
        'today'   : len([o for o in outs if o.get('date') == today_str])
    })

File: frontend/js/outbound.js
// ═══════════════════════════════════════════
// 불출: 품번 끝자리로 품목 검색
// ═══════════════════════════════════════════
async function searchBySuffix() {
  var suffix = document.getElementById('dp-suffix').value.trim();

  // 이전 결과 초기화
  document.getElementById('dp-step2').style.display = 'none';
  document.getElementById('dp-step3').style.display = 'none';
  selectedDpItem   = null;
  selectedFifoItems = [];

  if (suffix.length < 2) return;

  try {
    var res     = await fetch(API + '/fifo?code=' + encodeURIComponent(suffix));
    var json    = await res.json();
    var matches = json.data || [];

    if (!matches.length) {
      document.getElementById('dp-step2').style.display = 'block';
      document.getElementById('dp-item-list').innerHTML = '<p style="text-align:center;color:var(--txt2);padding:16px 0;font-size:14px">해당 품번 없음</p>';
      return;
    }

    // 품번 기준 중복 제거 (대표 항목만)
    var seen = {};
    var unique = [];
    matches.forEach(function(m) {
      if (m.code && !seen[m.code]) {
        seen[m.code] = true;
        unique.push(m);
      }
    });

    // 1개면 바로 3단계로
    if (unique.length === 1) {
      selectDpItem(unique[0]);
      return;
    }

    // 여러 개면 선택 목록 표시
    document.getElementById('dp-step2').style.display = 'block';
    document.getElementById('dp-item-list').innerHTML = unique.map(function(item) {
      return '<div class="dp-item-card" data-code="' + item.code + '" style="padding:12px;border:1px solid var(--border);border-radius:var(--r-md);margin-bottom:8px;cursor:pointer;background:var(--card);color:var(--txt)">' +
        '<div style="font-size:14px;font-weight:600;color:var(--txt)">' + (item.name || '-') + '</div>' +
        '<div style="font-size:12px;color:var(--txt2);margin-top:3px">' + item.code + '</div>' +
        '</div>';
    }).join('');

    // 품목 카드 클릭 이벤트
    document.querySelectorAll('.dp-item-card').forEach(function(card) {
      card.addEventListener('click', function() {
        var code = this.dataset.code;
        var found = unique.find(function(u){ return u.code === code; });
        if (found) selectDpItem(found);
      });
    });

  } catch(e) {
    showToast('서버 연결 오류');
  }
}


// ═══════════════════════════════════════════
// 불출: 품목 선택 → FIFO 목록 표시
// ═══════════════════════════════════════════
async function selectDpItem(item) {
  // 잠금 확인
  try {
    var lockRes = await fetch(API + '/lock/' + encodeURIComponent(item.code), { credentials: 'include' });
    var lockJson = await lockRes.json();
    if (lockJson.locked) {
      showToast('⚠️ ' + lockJson.name + ' 님이 작업 중입니다!');
      return;
    }
  } catch(e) {}
  // 잠금 등록
  fetch(API + '/lock', {
    method: 'POST', credentials: 'include',
    headers: {'Content-Type':'application/json'},
    body: JSON.stringify({ code: item.code })
  });

  selectedDpItem   = item;
  selectedFifoItems = [];

  document.getElementById('dp-step2').style.display = 'none';
  document.getElementById('dp-step3').style.display = 'block';

  // 선택된 품목 정보 표시
  document.getElementById('dp-selected-info').innerHTML =
    '<div style="font-size:13px;font-weight:600;color:var(--blue)">' + (item.name || '-') + '</div>' +
    '<div style="font-size:12px;color:var(--blue);margin-top:2px;opacity:.8">' + item.code + '</div>';

  // 품목 유형 자동 세팅
  setType('dp', (item.item_type === 'fabric' || item.itemType === 'fabric') ? 'fabric' : 'normal');

  // FIFO 순서 조회
  try {
    var res  = await fetch(API + '/fifo?code=' + encodeURIComponent(item.code), { credentials: 'include' });
    var json = await res.json();
    var list = json.data || [];

    var wrap = document.getElementById('dp-fifo-wrap');
    if (!list.length) { wrap.innerHTML = ''; return; }

    selectedFifoItems = [];

    // FIFO 목록 렌더링 (다중 선택)
    var rows = list.map(function(m, idx) {
      var qty  = m.item_type === 'fabric' ? (m.rolls + '롤/' + m.weight + 'kg' + (m.meters ? '/' + m.meters + 'm' : '')) : (m.qty + '개');
      var tag  = m.kind === 'hwanjip' ? '<span class="hj-tag">환입</span>' : '';
      var poTag = m.po ? ' · <span style="background:#fef9c3;color:#854d0e;padding:1px 5px;border-radius:4px;font-size:11px;font-weight:600">PO:' + m.po + '</span>' : '';
      return '<div class="fifo-row" data-id="' + m.id + '" style="display:flex;align-items:center;gap:8px">' +
        '<input type="checkbox" class="fifo-chk" data-id="' + m.id + '" style="width:18px;height:18px;cursor:pointer;flex-shrink:0">' +
        '<div class="fifo-n">' + (idx + 1) + '</div>' +
        '<div class="fifo-info" style="flex:1">' +
          '<div class="fifo-name">' + (m.name || m.code) + tag + '</div>' +
          '<div class="fifo-sub">📍 ' + (m.loc || '-') + ' · ' + qty + (m.lot ? ' · Lot:' + m.lot : '') + poTag + '</div>' +
        '</div>' +
        '<div class="fifo-dt">' + m.date + '</div>' +
      '</div>';
    }).join('');

    wrap.innerHTML =
      '<div class="fifo-box">' +
        '<div class="fifo-hd" style="display:flex;justify-content:space-between;align-items:center">' +
          '<span>선입선출 순서 — 체크해서 선택</span>' +
          '<span id="dp-selected-count" style="font-size:12px;color:var(--txt2)">0개 선택</span>' +
        '</div>' +
        rows +
      '</div>';

    // 체크박스 클릭 이벤트
    wrap.querySelectorAll('.fifo-chk').forEach(function(chk) {
      chk.addEventListener('change', function() {
        var id    = parseInt(this.dataset.id);
        var found = list.find(function(m){ return m.id === id; });
        if (!found) return;

        if (this.checked) {
          selectedFifoItems.push(found);
          this.closest('.fifo-row').classList.add('selected');
        } else {
          selectedFifoItems = selectedFifoItems.filter(function(m){ return m.id !== id; });
          this.closest('.fifo-row').classList.remove('selected');
        }

        // 선택 개수 표시
        var countEl = document.getElementById('dp-selected-count');
        if (countEl) countEl.textContent = selectedFifoItems.length + '개 선택';

        // 선택된 항목들 합계 자동 입력
        var totalQty = 0, totalRolls = 0, totalWeight = 0, totalMeters = 0;
        var isFabric = selectedFifoItems.length > 0 && selectedFifoItems[0].item_type === 'fabric';
        selectedFifoItems.forEach(function(m) {
          if (isFabric) { totalRolls += (m.rolls || 0); totalWeight += (m.weight || 0); totalMeters += (m.meters || 0); }
          else          { totalQty   += (m.qty   || 0); }
        });

        if (isFabric) {
          document.getElementById('dp-rolls').value   = totalRolls;
          document.getElementById('dp-weight').value  = totalWeight;
          document.getElementById('dp-meters').value  = totalMeters;
        } else {
          document.getElementById('dp-qty').value = totalQty;
        }

        // 선택한 항목이 있으면 3단계 표시
        if (selectedFifoItems.length > 0) {
          document.getElementById('dp-step3').style.display = 'block';
        }
      });
    });

    // 1번 자동 선택
    if (list.length) {
      wrap.querySelectorAll('.fifo-row')[0].click();
    }

  } catch(e) {
    showToast('FIFO 조회 오류');
  }
}


// ═══════════════════════════════════════════
// 불출: MAX 버튼 (선택된 항목 전량)
// ═══════════════════════════════════════════
function setMaxQty() {
  if (!selectedFifoItems || selectedFifoItems.length === 0) { showToast('먼저 항목을 선택하세요'); return; }
  var totalQty = 0, totalRolls = 0, totalWeight = 0, totalMeters = 0;
  selectedFifoItems.forEach(function(m) {
    totalQty    += (m.qty     || 0);
    totalRolls  += (m.rolls   || 0);
    totalWeight += (m.weight  || 0);
    totalMeters += (m.meters  || 0);
  });
  if (selectedFifoItems[0].item_type === 'fabric') {
    document.getElementById('dp-rolls').value   = totalRolls;
    document.getElementById('dp-weight').value  = totalWeight;
    document.getElementById('dp-meters').value  = totalMeters;
  } else {
    document.getElementById('dp-qty').value = totalQty;
  }
}


// ═══════════════════════════════════════════
// 불출 확정
// ═══════════════════════════════════════════
async function submitDispatch() {
  if (!selectedDpItem)   { showToast('품목을 먼저 선택하세요'); return; }
  if (!selectedFifoItems || selectedFifoItems.length === 0) { showToast('불출할 항목을 선택하세요'); return; }

  var person = document.getElementById('dp-person').value.trim();
  if (!person) { showToast('담당자 이름을 입력하세요'); return; }

  var qty = 0, rolls = 0, weight = 0, meters = 0;
  if (dpType === 'fabric') {
    rolls  = parseFloat(document.getElementById('dp-rolls').value)   || 0;
    weight = parseFloat(document.getElementById('dp-weight').value)  || 0;
    meters = parseFloat(document.getElementById('dp-meters').value)  || 0;
    qty    = rolls;
  } else {
    qty = parseFloat(document.getElementById('dp-qty').value) || 0;
  }
  if (!qty) { showToast('수량을 입력하세요'); return; }

  try {
    var res  = await fetch(API + '/outbound', {
      method: 'POST',
      credentials: 'include',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        item_type : dpType,
        code      : selectedDpItem.code,
        name      : selectedDpItem.name,
        person, qty, rolls, weight, meters,
        from_ids : selectedFifoItems.map(function(m){ return m.id; }),
        from_id  : selectedFifoItems[0].id,
        from_loc : selectedFifoItems.map(function(m){ return m.loc || '' }).join(', '),
        date     : document.getElementById('dp-date').value,
        note     : document.getElementById('dp-note').value,
      })
    });
    var json = await res.json();
    if (json.success) {
      showToast('✓ 불출 확정 완료!');
      if (selectedDpItem) fetch(API + '/lock/' + encodeURIComponent(selectedDpItem.code), { method: 'DELETE', credentials: 'include' });
      resetDispatch();
    } else {
      showToast('오류: ' + json.message);
    }
  } catch(e) {
    showToast('서버 연결 오류');
  }
}


// ═══════════════════════════════════════════
// 불출 초기화 (다시 선택)
// ═══════════════════════════════════════════
function resetDispatch() {
  if (selectedDpItem) fetch(API + '/lock/' + encodeURIComponent(selectedDpItem.code), { method: 'DELETE', credentials: 'include' });
  selectedDpItem   = null;
  selectedFifoItems = [];
  document.getElementById('dp-suffix').value = '';
  document.getElementById('dp-step2').style.display = 'none';
  document.getElementById('dp-step3').style.display = 'none';
  document.getElementById('dp-fifo-wrap').innerHTML = '';
  ['dp-qty','dp-rolls','dp-weight','dp-meters','dp-note'].forEach(function(id){
    var el = document.getElementById(id);
    if (el) el.value = '';
  });
  // 담당자는 초기화하지 않고 현재 로그인 사용자로 유지
  var dpPerson = document.getElementById('dp-person');
  if (dpPerson && currentUser) dpPerson.value = currentUser.name;
  setType('dp', 'normal');
}

